File: leetcode/python/1106.parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        i = 0
        def parse() -> bool:
            nonlocal i
            # print(expression[i:])
            if expression[i] == 't':
                i += 1
                return True
            if expression[i] == 'f':
                i += 1
                return False
            if expression[i] == '!':
                i += 2
                ans = parse()
                i += 1
                return not ans
            isAnd = True if expression[i] == '&' else False
            ans = isAnd
            i += 2
            while expression[i] != ')':
                parsedExpr = parse()
                if isAnd:
                    ans = ans and parsedExpr
                else:
                    ans = ans or parsedExpr
                if expression[i] == ',':
                    i += 1
            i += 1
            return ans
        return parse()

File: leetcode/rust-leetcode/src/1106.parsing-a-boolean-expression.rs
impl Solution {
    pub fn parse_bool_expr(expression: String) -> bool {
        Self::parse(&expression, &mut 0)
    }
    fn parse(expr: &String, idx: &mut usize) -> bool {
        let i = *idx;
        if &expr[i..i+1] == "t" {
            *idx += 1;
            return true;
        }
        if &expr[i..i+1] == "f" {
            *idx += 1;
            return false;
        }
        if &expr[i..i+1] == "!" {
            *idx += 2;
            let ans = !Self::parse(expr, idx);
            *idx += 1;
            return ans;
        }
        let isAnd = &expr[i..i+1] == "&";
        let mut ans = isAnd;
        *idx += 2;
        while &expr[(*idx)..(*idx+1)] != ")" {
            let parsed = Self::parse(expr, idx);
            if isAnd {
                ans &= parsed;
            } else {
                ans |= parsed;
            }
            if (&expr[(*idx)..(*idx+1)] == ",") {
                *idx += 1;
            }
        }
        *idx += 1;
        return ans;
    }
}

